lib/rules/layer-name-pattern/index.mjs

```diff
@@ -0,0 +1,88 @@
+import valueParser from 'postcss-value-parser';
+
+import { isRegExp, isString } from '../../utils/validateTypes.mjs';
+import { atRuleParamIndex } from '../../utils/nodeFieldIndices.mjs';
+import { atRuleRegexes } from '../../utils/regexes.mjs';
+import report from '../../utils/report.mjs';
+import ruleMessages from '../../utils/ruleMessages.mjs';
+import validateOptions from '../../utils/validateOptions.mjs';
+
+const ruleName = 'layer-name-pattern';
+
+const messages = ruleMessages(ruleName, {
+	expected: (name, pattern) => `Expected "${name}" to match pattern "${pattern}"`,
+});
+
+const meta = {
+	url: 'https://stylelint.io/user-guide/rules/layer-name-pattern',
+};
+
+/** @type {import('stylelint').CoreRules[ruleName]} */
+const rule = (primary) => {
+	return (root, result) => {
+		const validOptions = validateOptions(result, ruleName, {
+			actual: primary,
+			possible: [isRegExp, isString],
+		});
+
+		if (!validOptions) return;
+
+		const pattern = isString(primary) ? new RegExp(primary) : primary;
+
+		root.walkAtRules(atRuleRegexes.layerName, (atRule) => {
+			if (!atRule.params) return;
+
+			const parsedParams = valueParser(atRule.params);
+
+			parsedParams.walk((node) => {
+				if (node.type !== 'word') return;
+
+				if (pattern.test(node.value)) return;
+
+				const index = atRuleParamIndex(atRule) + node.sourceIndex;
+				const endIndex = index + node.value.length;
+
+				report({
+					message: messages.expected(node.value, primary),
+					node: atRule,
+					index,
+					endIndex,
+					ruleName,
+					result,
+				});
+			});
+		});
+
+		root.walkAtRules(atRuleRegexes.importName, (atRule) => {
+			const parsedParams = valueParser(atRule.params);
```

Comment:
```suggestion
            const { params } = atRule;
            if (!LAYER_FUNC_REGEX.test(params)) return
			const parsedParams = valueParser(params);
```

```js
// Near top of file:
const LAYER_FUNC_REGEX = /layer\(/i;
```

When possible, we typically do a cheap check before parsing values.

Reply:
Yes, good point. 👍🏼 

We, especially Romain, have worked on improving rule performance. Avoiding extra parsing is worthwhile. See https://github.com/stylelint/stylelint/issues/6869